docs/RFCS/20210330_sql_stats_persistence.md

```diff
@@ -0,0 +1,407 @@
+      particular node.
+   1. Due to the reliance on the RPC-fanout, post-processings of the SQL
+      statistics (e.g. sorting, filtering) are currently implemented within DB
+      console. As we move toward storing and displaying historical statistics,
+      solely relying on DB console to perform slicing-n-dicing of the
+      statistics data is not scalable.
+1. As CRDB moving toward multi-tenant architecture, relying on fanout RPC
+   implies that tenant statistics aggregation will not only have the drawbacks
+   mentioned previously, but also be depending on the progress of pod-to-pod
+   communication implementation, which is not ideal.
+
+The persistence of SQL statistics in a CRDB system table can addresses existing
+drawbacks. CRDB will gain improvement in two areas:
+1. **Usability**: with DistSQL we will be able to process more complex queries
+  to answer the questions users might have for the performance of their queries
+  overtime.
+1. **Reliability**: with CRDB SQL statistics now backed by a persistent table,
+  we will ensure the survival of the data across node crash/upgrade/restarts.
+
+# Design
+
+## Design Considerations
+
+* Collected SQL statistics need to be available on every node that receives SQL
+  queries and the accumulated statistics need to survive node restart/crash.
+
+* Collected statistics should be able to answer users' potential questions for
+  their queries overtime through both DB Console and SQL shell.
+
+* Statistics persistence should be low overhead, but the collected statistics
+  should also enough resolution to provide meaningful insight into the
+  query/txn performance.
+
+* There is a need for mechanism to prune old statistics data to reduce the
+  burden on storage space.
+
+* Statistics collection and statistics persistence should be decoupled.
+
+## Design Overview
+
+Two new system tables `system.experimetnal_sql_stmt_stats` and
+`system.experimental_sql_txn_stats` provide storage for storing time series
+data for accumulated statistics for statements and transactions.
+
+Currently, each CRDB node stores in-memory statistics for transactions and
+statements in which the node is the gateway for. The in-memory statistics is
+flushed into system tables in one of the following scenarios:
+1. at the end of a fixed interval determined by a cluster setting. E.g. every 5
+   mins.
+1. when user explicitly requests all in-memory statistics to be persisted. E.g.
+   through a SQL shell builtin or through DB Console.
+
+During the flush operation, for each statement/transaction fingerprint, the
+CRDB node will check if there already exists same fingerprint in the persisted
+system tables within the latest time bucket.
+* if such entry exists, the flush operation will aggregate the existing entry.
+* if such extry does not exist, the flush operation will insert a new entry.
+
+When DB Console issues fetch requests to CRDB node through HTTP endpoint, the
```

Comment:
Do the two tables (sql_stmt_stats and sql_txn_stats) provide access **only** to the persisted statistics? Have we considered extending this view to represent in-memory statistics as well where when the DB Console (or developer) queries the tables(s), we transparently do a union and/or aggregate across the current in-memory statistics and persisted statistics?